src/track/beats.cpp

```diff
@@ -53,3 +190,676 @@ double Beats::findNBeatsFromSample(double fromSample, double beats) const {
+    case HALVE:
+        // remove every second beat
+        scaleHalve();
+        break;
+    case TWOTHIRDS:
+        // introduce a new beat into every gap
+        scaleDouble();
+        // remove every second and third beat
+        scaleThird();
+        break;
+    case THREEFOURTHS:
+        // introduce two beats into every gap
+        scaleTriple();
+        // remove every second third and forth beat
+        scaleFourth();
+        break;
+    case FOURTHIRDS:
+        // introduce three beats into every gap
+        scaleQuadruple();
+        // remove every second third and forth beat
+        scaleThird();
+        break;
+    case THREEHALVES:
+        // introduce two beats into every gap
+        scaleTriple();
+        // remove every second beat
+        scaleHalve();
+        break;
+    default:
+        DEBUG_ASSERT(!"scale value invalid");
+        return;
+    }
+    onBeatlistChanged();
+    locker.unlock();
+    emit updated();
+}
+
+void Beats::scaleDouble() {
+    track::io::Beat prevBeat = m_beats.first();
+    // Skip the first beat to preserve the first beat in a measure
+    BeatList::iterator it = m_beats.begin() + 1;
+    for (; it != m_beats.end(); ++it) {
+        // Need to not accrue fractional frames.
+        int distance = it->frame_position() - prevBeat.frame_position();
+        track::io::Beat beat;
+        beat.set_frame_position(prevBeat.frame_position() + distance / 2);
+        it = m_beats.insert(it, beat);
+        prevBeat = (++it)[0];
+    }
+}
+
+void Beats::scaleTriple() {
+    track::io::Beat prevBeat = m_beats.first();
+    // Skip the first beat to preserve the first beat in a measure
+    BeatList::iterator it = m_beats.begin() + 1;
+    for (; it != m_beats.end(); ++it) {
+        // Need to not accrue fractional frames.
+        int distance = it->frame_position() - prevBeat.frame_position();
+        track::io::Beat beat;
+        beat.set_frame_position(prevBeat.frame_position() + distance / 3);
+        it = m_beats.insert(it, beat);
+        ++it;
+        beat.set_frame_position(prevBeat.frame_position() + distance * 2 / 3);
+        it = m_beats.insert(it, beat);
+        prevBeat = (++it)[0];
+    }
+}
+
+void Beats::scaleQuadruple() {
+    track::io::Beat prevBeat = m_beats.first();
+    // Skip the first beat to preserve the first beat in a measure
+    BeatList::iterator it = m_beats.begin() + 1;
+    for (; it != m_beats.end(); ++it) {
+        // Need to not accrue fractional frames.
+        int distance = it->frame_position() - prevBeat.frame_position();
+        track::io::Beat beat;
+        for (int i = 1; i <= 3; i++) {
+            beat.set_frame_position(prevBeat.frame_position() + distance * i / 4);
+            it = m_beats.insert(it, beat);
+            ++it;
+        }
+        prevBeat = it[0];
+    }
+}
+
+void Beats::scaleHalve() {
+    // Skip the first beat to preserve the first beat in a measure
+    BeatList::iterator it = m_beats.begin() + 1;
```

Comment:
This could be written more concise:
```
auto it = m_beats.begin();
while (it != m_beats.end) {
    // Keep every 1st beat (including the 1st beat in a measure)
    ++it;
    if (it == m_beats.end) {
        break;
    }
    // Drop every 2nd beat
    it = m_beats.erase(it);
}
```